File: algorithms/code/leetcode/lc005_longest_palindromic_substring/LC005LongestPalindromicSubstring.cs
// https://leetcode.com/problems/longest-palindromic-substring/
// See also: Manacher algorithm

namespace AlgorithmDesign.code.leetcode.lc005_longest_palindromic_substring
{
    public class Solution
    {
        public string LongestPalindrome(string s)
        {
            string longest = "";
            int index = 0;
            while (index < s.Length)
            {
                // Find head & tail
                int head = index, tail = index;
                while (tail + 1 < s.Length)
                {
                    if (s[head] == s[tail + 1])
                    {
                        tail++;
                    }
                    else
                    {
                        break;
                    }
                }
                index += 1 + tail - head;
                // Expand
                while (0 <= head - 1 && tail + 1 < s.Length && s[head - 1] == s[tail + 1])
                {
                    head--;
                    tail++;
                }
                // Check
                if (longest.Length < 1 + tail - head)
                {
                    longest = s.Substring(head, 1 + tail - head);
                }
            }
            return longest;
        }
    }
}

File: algorithms/code/leetcode/lc005_longest_palindromic_substring/lc005_longest_palindromic_substring.go
// Package lc005 implements https://leetcode.com/problems/longest-palindromic-substring/
// See also: Manacher algorithm
package lc005

func longestPalindrome(s string) string {
	longest := ""
	n := len(s)
	for index := 0; index < n; {
		// Find head & tail
		head, tail := index, index
		for tail+1 < n && s[head] == s[tail+1] {
			tail++
		}
		index += 1 + tail - head
		// Expand
		for 0 <= head-1 && tail+1 < n && s[head-1] == s[tail+1] {
			head--
			tail++
		}
		// Check
		if len(longest) < 1+tail-head {
			longest = s[head : 1+tail]
		}
	}
	return longest
}
